README.md
A Firefox extension that shows you replies to all of your 
[Hacker News](https://news.ycombinator.com) posts and comments within the past
week. Refreshes every 10 minutes.

![Extension screenshot](https://imgur.com/3UwZYEJ.png)


CHANGELOG.md
## [1.0.1] - 07/30/19
### Added
- Show full timestamp when hovering on reply's time label
### Changed
- Minor visual spacing tweak
### Fixed
- Clicking 'dismiss' would scroll to top

[Unreleased]: https://github.com/mateoeh/paperboy/compare/v1.0.1...HEAD
[1.0.1]: https://github.com/mateoeh/paperboy/releases/tag/v1.0.1


src/js/common.js
// FIXME
// eslint-disable-next-line no-var, no-use-before-define
var paperboy = paperboy || {};

paperboy.Reply = function Reply(postId, author, timestamp, content) {
    this.postId = postId;
    this.author = author;
    this.timestamp = timestamp;
    this.content = content;
};

// Constants
paperboy.STORAGE = browser.storage.local;
// Message types
paperboy.REFRESH = "refresh";
paperboy.REFRESHED = "refreshed";
paperboy.NETWORK_ERROR = "network_error";
paperboy.NO_SUCH_USER = "no_such_user";
paperboy.BADGE_UPDATE = "badge_update";

/**
 * Sends a message to all background scripts
 * @param {Object} msg - any Object. `msg.event` should be `paperboy.{REFRESH,
 * REFRESHED, NETWORK_ERROR, NO_SUCH_USER, BADGE_UPDATE}`
 */
paperboy.send = function sendMessage(msg) {
    browser.runtime.sendMessage(msg);
};

/**
 * Generates a "time ago" string for a given timestamp e.g. '2 minutes ago'
 *
 * @param {number} unixTime - Seconds since Unix epoch
 * @return {string} Time ago string
 */
/* eslint-disable nonblock-statement-body-position,curly */
paperboy.timeAgoString = function timeAgoString(unixTime) {
    const { floor } = Math;
    const now = floor(Date.now() / 1000);
    const diff = now - unixTime;

    if (diff < 2)
        // 00:00:02
        return "just now";
    if (diff < 60)
        // 00:01:00
        return `${diff} seconds ago`;
    if (diff < 60 * 2)
        // 00:02:00
        return "1 minute ago";
    if (diff < 60 * 60)
        // 01:00:00
        return `${floor(diff / 60)} minutes ago`;
    if (diff < 60 * 60 * 2)
        // 02:00:00
        return "1 hour ago";
    if (diff < 60 * 60 * 24)
        // 24:00:00
        return `${floor(diff / 60 / 60)} hours ago`;
    if (diff < 60 * 60 * 24 * 2)
        // 48:00:00
        return "1 day ago";
    if (diff < 60 * 60 * 24 * 7)
        // 7 days
        return `${floor(diff / 60 / 60 / 24)} days ago`;
    if (diff < 60 * 60 * 24 * 7 * 2)
        // 14 days
        return "1 week ago";
    if (diff < 60 * 60 * 24 * 30.44)
        // 30.44 days
        return `${floor(diff / 60 / 60 / 24 / 7)} weeks ago`;
    if (diff < 60 * 60 * 24 * 30.44 * 2)
        // 60.88 days
        return "1 month ago";
    if (diff < 60 * 60 * 24 * 30.44 * 12)
        // 1 year
        return `${floor(diff / 60 / 60 / 24 / 30.44)} months ago`;
    if (diff < 60 * 60 * 24 * 30.44 * 12 * 2)
        // 2 years
        return "1 year ago";
    return `${floor(diff / 60 / 60 / 24 / 30.44 / 12)} years ago`;
};
/* eslint-enable */


src/popup/popup.js
const body = document.querySelector("body");
const usernameForm = document.querySelector("form");
const usernameInput = document.querySelector("#username-input");
const lastRefreshSpan = document.querySelector("#last-refresh");
const clearAllButton = document.querySelector("#clear-all");
const replyTemplate = document.querySelector("template");
const emptyLabel = document.querySelector("#no-notifications");

let repliesCache = null;
let usernameCache = null;

const removeAllReplyElements = function removeAllReplyElements() {
    const divs = body.getElementsByClassName("reply-container");
    while (divs[0]) {
        body.removeChild(divs[0]);
    }
};

const showEmptyState = function showEmptyState(isEmpty) {
    emptyLabel.style.display = isEmpty ? "inline-block" : "none";
    clearAllButton.style.visibility = isEmpty ? "hidden" : "visible";
};

const clearReply = function clearReply(clickEvent) {
    const { postId } = clickEvent.target;
    repliesCache[usernameCache] = repliesCache[usernameCache].filter(
        reply => reply.postId !== postId
    );
    paperboy.send({
        event: paperboy.BADGE_UPDATE,
        count: repliesCache[usernameCache].length
    });
    showEmptyState(repliesCache[usernameCache].length === 0);
    body.removeChild(clickEvent.target.parentElement);
    paperboy.STORAGE.set({ replies: repliesCache });
};

const clearAll = function clearAll(clickEvent) {
    clickEvent.preventDefault();
    repliesCache[usernameCache] = [];
    paperboy.send({
        event: paperboy.BADGE_UPDATE,
        count: 0
    });
    showEmptyState(true);
    removeAllReplyElements();
    paperboy.STORAGE.set({ replies: repliesCache });
};

const bindReplyToContainer = function bindReplyToContainer(reply, container) {
    const link = container.querySelector(".item-link");
    link.href = `https://news.ycombinator.com/item?id=${reply.postId}`;
    link.textContent = `Reply from ${reply.author}`;
    link.postId = reply.postId;
    link.addEventListener("click", clearReply);

    const time = container.querySelector(".time-label");
    time.textContent = paperboy.timeAgoString(reply.timestamp);
    time.title = new Date(reply.timestamp * 1000).toString();

    const content = container.querySelector(".content");
    content.textContent = reply.content;

    const clear = container.querySelector(".clear");
    clear.postId = reply.postId;
    clear.addEventListener("click", clickEvent => {
        clickEvent.preventDefault();
        clearReply(clickEvent);
    });
};

const showRepliesForUser = function showRepliesForUser(replies, username) {
    repliesCache = replies;
    usernameCache = username;

    for (let i = 0; i < replies[username].length; i += 1) {
        const container = document.importNode(replyTemplate.content, true);
        bindReplyToContainer(replies[username][i], container);
        body.appendChild(container);
    }

    showEmptyState(replies[username].length === 0);
};

const didRefresh = async function didRefresh() {
    const { username, replies, lastRefresh } = await paperboy.STORAGE.get([
        "username",
        "replies",
        "lastRefresh"
    ]);

    if (
        typeof username === "undefined" ||
        typeof replies === "undefined" ||
        typeof replies[username] === "undefined"
    ) {
        showEmptyState(true);
        paperboy.send({
            event: paperboy.BADGE_UPDATE,
            count: 0
        });
    } else {
        usernameInput.value = username;
        lastRefreshSpan.textContent = `Last refresh: ${paperboy.timeAgoString(
            lastRefresh[username]
        )}`;
        removeAllReplyElements();
        showRepliesForUser(replies, username);
    }
};

const saveUsername = function saveUsername(clickEvent) {
    clickEvent.preventDefault();
    lastRefreshSpan.textContent = "Refreshing...";
    paperboy.STORAGE.set({ username: usernameInput.value });
    paperboy.send({ event: paperboy.REFRESH });
};

window.addEventListener("DOMContentLoaded", didRefresh);
usernameForm.addEventListener("submit", saveUsername);
clearAllButton.addEventListener("click", clearAll);

browser.runtime.onMessage.addListener(msg => {
    switch (msg.event) {
        case paperboy.REFRESHED:
            didRefresh();
            break;
        case paperboy.NETWORK_ERROR:
            lastRefreshSpan.textContent = "Network error";
            break;
        case paperboy.NO_SUCH_USER:
            lastRefreshSpan.textContent = "User doesn't exist";
            paperboy.send({
                event: paperboy.BADGE_UPDATE,
                count: 0
            });
            removeAllReplyElements();
            showEmptyState(true);
            break;
        default:
            // Ignore
    }
});


src/js/background.js
const BASE_URL = "https://hacker-news.firebaseio.com/v0";

class NoSuchUserError extends Error {
    constructor(message) {
        super(message);
        this.name = "NoSuchUserError";
    }
}

const hasOwnProperty = function hasOwnProperty(object, property) {
    return Object.prototype.hasOwnProperty.call(object, property);
};

const deepCopy = function deepCopy(obj) {
    return JSON.parse(JSON.stringify(obj));
};

const setBadgeCount = function setBadgeCount(count) {
    if (count === 0) {
        browser.browserAction.setBadgeText({ text: "" });
    } else {
        browser.browserAction.setBadgeText({ text: count.toString() });
    }
};

let decodingTextArea = null;
// Make HTML suitable for .textContent
const toPlaintext = function toPlaintext(html) {
    // Strip HTML tags
    const sanitized = html.replace(/<p>/g, " ").replace(/<[^>]*>/g, "");

    // Decode HTML entities. <textarea> will preserve tags (to avoid XSS)
    // https://stackoverflow.com/a/7394787
    if (decodingTextArea === null) {
        decodingTextArea = document.createElement("textarea");
    }
    decodingTextArea.innerHTML = sanitized;
    return decodingTextArea.value;
};

const getUserJson = function getUserJson(username) {
    return fetch(`${BASE_URL}/user/${username}.json`)
        .then(res => res.json())
        .then(json => {
            if (json === null) throw new NoSuchUserError();
            else return json;
        });
};

// Get all of user's posts within the last week
const getRecentPosts = async function getRecentPosts(userJson) {
    const now = Math.floor(Date.now() / 1000);
    const recents = [];

    for (let i = 0; i < userJson.submitted.length; i += 1) {
        const postId = userJson.submitted[i];

        // Run these sequentially so loop can break early
        /* eslint-disable no-await-in-loop */
        const post = await fetch(`${BASE_URL}/item/${postId}.json`);
        const json = await post.json();
        /* eslint-enable */

        if (json.time > now - 60 * 60 * 24 * 7) {
            recents.push(json);
        } else {
            break;
        }
    }

    return recents;
};

const getChildren = function getChildren(recentJsons, after) {
    const childIds = recentJsons
        .filter(item => hasOwnProperty(item, "kids"))
        .flatMap(item => item.kids)
        .filter(id => id > after)
        .sort((a, b) => b - a); // high to low
    return Promise.all(
        childIds.map(id =>
            fetch(`${BASE_URL}/item/${id}.json`).then(res => res.json())
        )
    );
};

const toReplies = function toReplies(childJsons) {
    return childJsons
        .filter(json => !hasOwnProperty(json, "deleted"))
        .filter(json => !hasOwnProperty(json, "dead"))
        .map(
            json =>
                new paperboy.Reply(
                    json.id,
                    json.by,
                    json.time,
                    toPlaintext(json.text)
                )
        );
};

/* <storage convenience methods> */
const storeLastRefresh = function storeLastRefresh(
    username,
    lastRefreshCache = {}
) {
    const lastRefreshCopy = deepCopy(lastRefreshCache);
    lastRefreshCopy[username] = Math.floor(Date.now() / 1000);
    return paperboy.STORAGE.set({ lastRefresh: lastRefreshCopy });
};

const storeReplies = function storeReplies(
    username,
    fetchedReplies,
    repliesCache = {}
) {
    const repliesCopy = deepCopy(repliesCache);
    repliesCopy[username] = repliesCache[username] || [];
    repliesCopy[username] = fetchedReplies.concat(repliesCopy[username]);
    setBadgeCount(repliesCopy[username].length);
    return paperboy.STORAGE.set({ replies: repliesCopy });
};

const storeHighestSeen = function storeHighestSeen(
    username,
    fetchedReplies,
    highestSeenCache = {}
) {
    if (fetchedReplies.length === 0) return null;
    const highestSeenCopy = deepCopy(highestSeenCache);
    highestSeenCopy[username] = fetchedReplies[0].postId;
    return paperboy.STORAGE.set({ highestSeen: highestSeenCopy });
};
/* </storage convenience methods> */

const handleErrors = function handleErrors(error) {
    console.log(`Error during refresh: ${error}`);
    if (error instanceof NoSuchUserError) {
        paperboy.send({ event: paperboy.NO_SUCH_USER });
    } else {
        paperboy.send({ event: paperboy.NETWORK_ERROR });
    }
};

const refresh = async function refresh() {
    const {
        username,
        replies,
        lastRefresh,
        highestSeen
    } = await paperboy.STORAGE.get([
        "username",
        "replies",
        "lastRefresh",
        "highestSeen"
    ]);

    if (typeof username === "undefined") {
        // No username set, nothing to fetch
        return;
    }

    try {
        setBadgeCount(replies[username].length);
    } catch (err) {
        // Something is undefined. Ignore
    }

    let minimumId;
    if (
        typeof highestSeen !== "undefined" &&
        typeof highestSeen[username] !== "undefined"
    ) {
        // Don't fetch previously seen replies
        minimumId = highestSeen[username];
    } else {
        // First run for this user, fetch all replies
        minimumId = 0;
    }

    getUserJson(username)
        .then(getRecentPosts)
        .then(recentJsons => getChildren(recentJsons, minimumId))
        .then(toReplies)
        .then(fetchedReplies =>
            Promise.all([
                storeHighestSeen(username, fetchedReplies, highestSeen),
                storeReplies(username, fetchedReplies, replies),
                storeLastRefresh(username, lastRefresh)
            ])
        )
        .then(() => paperboy.send({ event: paperboy.REFRESHED }))
        .catch(handleErrors);
};

browser.browserAction.setBadgeBackgroundColor({ "color": "#ff6600" });
window.setInterval(refresh, 10 * 60 * 1000);
refresh();

browser.runtime.onMessage.addListener(msg => {
    if (msg.event === paperboy.REFRESH) {
        refresh();
    } else if (msg.event === paperboy.BADGE_UPDATE) {
        setBadgeCount(msg.count);
    }
});
